lib/runBlock.js

```diff
@@ -32,237 +32,187 @@ module.exports = function (opts, cb) {
-      }
-      return
+    const txReceipt = {
+      status: txRes.vm.exception ? 1 : 0, // result.vm.exception is 0 when an exception occurs, and 1 when it doesn't.  TODO make this the opposite
+      gasUsed: gasUsed.toArrayLike(Buffer),
+      bitvector: txRes.bloom.bitvector,
+      logs: txRes.vm.logs || []
     }
+    receipts.push(txReceipt)
 
-    self.stateManager.commit(function (err) {
-      if (err) return cb(err)
+    // Add receipt to trie to later calculate receipt root
+    await promisify(receiptTrie.put).bind(receiptTrie)(rlp.encode(txIdx), rlp.encode(Object.values(txReceipt)))
+  }
 
-      self.stateManager.getStateRoot(function (err, stateRoot) {
-        if (err) return cb(err)
+  return { bloom, gasUsed, receiptRoot: receiptTrie.root, receipts, results: txResults }
+}
 
-        // credit all block rewards
-        if (generateStateRoot) {
-          block.header.stateRoot = stateRoot
-        }
+/**
+ * Calculates block rewards for miner and ommers and puts
+ * the updated balances of their accounts to state.
+ */
+async function assignBlockRewards (block) {
+  const state = new PStateManager(this.stateManager)
+  const minerReward = new BN(this._common.param('pow', 'minerReward'))
+  const ommers = block.uncleHeaders
+  // Reward ommers
+  for (let ommer of ommers) {
+    const reward = calculateOmmerReward(ommer, block.header.number, minerReward)
+    await rewardAccount(state, ommer.coinbase, reward)
+  }
+  // Reward miner
+  const reward = calculateMinerReward(minerReward, ommers)
+  await rewardAccount(state, block.header.coinbase, reward)
+}
 
-        if (validateStateRoot) {
-          if (receiptTrie.root && receiptTrie.root.toString('hex') !== block.header.receiptTrie.toString('hex')) {
-            err = new Error((err || '') + 'invalid receiptTrie ')
-          }
-          if (bloom.bitvector.toString('hex') !== block.header.bloom.toString('hex')) {
-            err = new Error((err || '') + 'invalid bloom ')
-          }
-          if (ethUtil.bufferToInt(block.header.gasUsed) !== Number(gasUsed)) {
-            err = new Error((err || '') + 'invalid gasUsed ')
-          }
-          if (stateRoot.toString('hex') !== block.header.stateRoot.toString('hex')) {
-            err = new Error((err || '') + 'invalid block stateRoot ')
-          }
-        }
+function calculateOmmerReward (ommer, blockNumber, minerReward) {
+  const heightDiff = new BN(blockNumber).sub(new BN(ommer.number))
+  let reward = ((new BN(8)).sub(heightDiff)).mul(minerReward.divn(8))
+  if (reward.ltn(0)) {
+    reward = new BN(0)
+  }
+  return reward
+}
 
-        result = {
-          receipts: receipts,
-          results: txResults,
-          error: err
-        }
+function calculateMinerReward (minerReward, ommers) {
+  // calculate nibling reward
+  const niblingReward = minerReward.divn(32)
+  const totalNiblingReward = niblingReward.muln(ommers.length)
+  const reward = minerReward.add(totalNiblingReward)
+  return reward
+}
```

Comment:
Some cleanup on miner/ommer (side note: we should rename to "uncle", this confused me for quite some time) code, looks good.